src/apps/wall/scanner.lua

```diff
@@ -0,0 +1,299 @@
+--
+local function ipv6_nexthdr_type_len_skip (p)
+   return p[0], p + 8 + (p[1] * 8)
+end
+
+local ipv6_walk_header_funcs = {
+   [IPv6_NEXTHDR_HOPBYHOP] = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_ROUTING]  = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_DSTOPTS]  = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_FRAGMENT] = function (p)
+      return p[0], p + 8
+   end;
+   [IPv6_NEXTHDR_AH] = function (p)
+      -- Size specified in 4-octet units (plus two octets).
+      return p[0], p + 2 + (p[1] * 4)
+   end;
+}
+
+--
+-- Traverses all the IPv6 headers (using the "next header" fields) until an
+-- upper-level protocol header (e.g. TCP, UDP) is found. The returned value
+-- is the type of the upper level protocol code and pointer to the beginning
+-- of the upper level protocol header data.
+--
+local function ipv6_walk_headers (p)
+   local ptr = p.data
+   local nexthdr = ptr[IPv6_NEXTHDR_OFFSET]
+   while ipv6_walk_header_funcs[nexthdr] do
+      local new_nexthdr, new_ptr = ipv6_walk_header_funcs[nexthdr](ptr)
+      if new_ptr > p.data + p.length then
+         break
+      end
+      nexthdr, ptr = new_nexthdr, new_ptr
+   end
+   return nexthdr, ptr
+end
+
+
+Scanner = subClass()
+Scanner._name = "SnabbWall base packet Scanner"
+
+function Scanner:extract_packet_info(p)
+   local eth_type  = rd16(p.data + ETH_TYPE_OFFSET)
+   local ip_offset = ETH_HEADER_SIZE
+   local vlan_id   = 0
+
+   while eth_type == ETH_TYPE_VLAN do
+      vlan_id   = rd16(p.data + ip_offset)
+      eth_type  = rd16(p.data + ip_offset + 2)
+      ip_offset = ip_offset + 4
+   end
+
+   local key, src_addr, src_port, dst_addr, dst_port, ip_proto
+   if eth_type == ETH_TYPE_IPv4 then
+      key = flow_key_ipv4()
+      src_addr = rd32(p.data + ip_offset + IPv4_SRC_ADDR_OFFSET)
+      dst_addr = rd32(p.data + ip_offset + IPv4_DST_ADDR_OFFSET)
+      if src_addr < dst_addr then
+         key.lo_addr, key.hi_addr = src_addr, dst_addr
+      else
+         key.lo_addr, key.hi_addr = dst_addr, src_addr
+      end
+
+      ip_proto = p.data[ip_offset + IPv4_PROTO_OFFSET]
+      local ip_payload_offset = ip_offset + ihl(p, ip_offset)
+      if ip_proto == IPv4_PROTO_TCP then
+         src_port = rd16(p.data + ip_payload_offset + TCP_SRC_PORT_OFFSET)
+         dst_port = rd16(p.data + ip_payload_offset + TCP_DST_PORT_OFFSET)
+      elseif ip_proto == IPv4_PROTO_UDP then
+         src_port = rd16(p.data + ip_payload_offset + UDP_SRC_PORT_OFFSET)
+         dst_port = rd16(p.data + ip_payload_offset + UDP_DST_PORT_OFFSET)
+      end
+   elseif eth_type == ETH_TYPE_IPv6 then
+      key = flow_key_ipv6()
+      src_addr = p.data + IPv6_SRC_ADDR_OFFSET
+      dst_addr = p.data + IPv6_DST_ADDR_OFFSET
+      if ipv6_addr_cmp(src_addr, dst_addr) <= 0 then
+         ffi.copy(key:lo_addr_ptr(), src_addr, 16)
+         ffi.copy(key:hi_addr_ptr(), dst_addr, 16)
+      else
+         ffi.copy(key:lo_addr_ptr(), dst_addr, 16)
+         ffi.copy(key:hi_addr_ptr(), dst_addr, 16)
```

Comment:
I think it should be `src_addr` in this line instead of `dst_addr` (potential copy-paste error?).